# 4.6 Combining & Exporting Data Part 2

In [1]:
# Import Libraries

import pandas as pd
import numpy as np
import os

## Import df_orders_products_combined from PKL file

In [2]:
# Import df_orders_products_combined from PKL file

path = r'C:\Users\natha\OneDrive\Desktop\Data Analytics\Jupyter\09-2023 Instacart Basket Analysis'
df_orders_products_combined = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))


In [3]:
# Check output

df_orders_products_combined.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  is_first_order  product_id  add_to_cart_order  \
0                     NaN               1         196                  1   
1                     NaN               1       14084                  2   
2                     NaN               1       12427                  3   
3                     NaN               1       26088                  4   
4                     NaN               1       26405                  5   

   reordered _merge  
0          0   both  
1          0   both  
2          0   both  
3          0   both  
4          0   both

In [4]:
# Check shape

df_orders_products_combined.shape

(32434489, 11)

## Import the 'checked' df_prods

In [5]:
# Import 'checked' df_prods from products_checked.csv

path = r'C:\Users\natha\OneDrive\Desktop\Data Analytics\Jupyter\09-2023 Instacart Basket Analysis'
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index_col = False)

In [6]:
# Check output

df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [7]:
# Check shape

df_prods.shape

(49671, 5)

# Merge 'df_prods' and 'df_orders_products_combined' based on common column 'product_id' (INNER JOIN)
## New dataframe will be 'df_orders_products_merged'

## *Before starting the merge of the two dataframes, we must remove the '_merge' column that exists in 'df_orders_products_combined'

In [9]:
# Remove the _merge Column from df_orders_products_combined:

# Since the _merge column is not relevant for our current analysis, you can drop it before performing the new merge:

df_orders_products_combined = df_orders_products_combined.drop(columns=['_merge'])

In [10]:
# Merge 'df_prods' and 'df_orders_products_combined' based on common column 'product_id' as the 'key'

df_orders_products_merged = df_prods.merge(df_orders_products_combined, on = 'product_id', indicator = True)

In [11]:
# check merge flag frequency (inner joins) [inner joins labelled as 'both']

df_orders_products_merged['_merge'].value_counts()

_merge
both          32404161
left_only            0
right_only           0
Name: count, dtype: int64

In [12]:
# check merged dataframe (inner joins)

df_orders_products_merged.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  is_first_order  add_to_cart_order  reordered _merge  
0                     3.0               0                  5          0   both  
1                    20.0               0                  1          1   both  
2                     6.0               0                 20          0   both  
3                     NaN               1                 10          0   both  
4                     9.0               0                 11          1   both

### *Inner Join merge is what we want for our Instacart Project, as we want merges with a full match for our purposes. In this case we have 32404161 rows with full merges.*

# Exporting Data to CSV and PKL files

In [13]:
# Export df_orders_products_merged dataframe to CSV

df_orders_products_merged.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged.csv'), index=False)

In [15]:
# Export df_orders_products_merged dataframe to PKL

df_orders_products_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged.pkl'))